### PR TITLE
Canon integration

### PR DESCRIPTION
**Pull request by krl**
Current status, library builds, tests do not yet build.

Call functionality still commented out, but basic replacements are done.

### CODE REVIEW

**Comment by krl** [comment]
Still blocked on the new canonical release, will remove the patches in Cargo.toml when published.

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:59]: If we are cloning, then we should receive `callee` by value, and not reference

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:121]: The function returns a result but we opted to panic. Why not return `Err` instead?

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:146]: Same as above; maybe return `Err`?

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:181]: Same as above; maybe return `Err`?

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:206]: Same as above; maybe return `Err`?

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:215]: Same as above; maybe return `Err`?

File: src/call_context.rs
Comment (vlopes11) [inline src/call_context.rs:235]: Same as above; maybe return `Err`?

File: src/lib.rs
Comment (vlopes11) [inline src/lib.rs:31]: The proof verification failure don't seem to be contemplated in the error variants

File: src/state.rs
Comment (vlopes11) [inline src/state.rs:23]: Zedger trees will not have `17` depth. We need to think on a solution for trees of different size (maybe upgrade the const definition of the size to the network state itself?)

Also, by definition, contracts may have smaller poseidon trees, if they opt to

**Review by vlopes11 — COMMENTED** [review]
Some nits to discuss

**Review by vlopes11 — APPROVED** [review]
LGTM; small nits, find the best strategy to address them

**Review by ZER0 — CHANGES_REQUESTED** [review]
The code looks good, but there are few changes needs to be done, or at least things that needs to be clarify; I have some concerns about part of the code that is removed (e.g. injecting gas, floating numbers, etc).

But, most importantly, the commit history looks messy. I cannot understand from the commit message what it is about. This clearly cannot merged as is, the easiest solution would be squash the commit, and add one single commit that list all the notable changes; what was added, what was removed, why etc – not "how", that's the code for.

You can use those guidelines: https://chris.beams.io/posts/git-commit/

If you prefer avoid squashing it's fine to me, but it means you have to rewrite every single commit message following the above guidelines and I think it would take more time, but it's up to you.

Also, it looks like – at least locally – that there is some formatting error when I run `cargo fmt --all -- --check`. Currently our CI might not catch that since it's outdated, but it would makes more sense to move from Travis to Github Actions once this PR is landed.

File: Cargo.toml
Comment (krl) [reply Cargo.toml]: Removed the deprecated dependencies.

File: src/contract.rs
Comment (krl) [reply src/contract.rs:26]: Insure compiled was probably premature optimization.

For the other things, we will have to port it again, it will also have to be part of how we do the genesis setup, as per before, to be able to get the contract id before the gas injection code. I will file a new issue on this topic.

File: src/state.rs
Comment (krl) [reply src/state.rs:23]: This will be resolved by the removal of const arrayvec in microkelvin.

File: src/lib.rs
Comment (krl) [reply src/lib.rs:31]: We will adapt the error type when we need it.

File: dusk-abi/src/lib.rs
Comment (krl) [reply dusk-abi/src/lib.rs:316]: The user should be using the debug!() macro, hiding here is the easier way to keep everything in scope.

File: src/call_context.rs
Comment (krl) [reply src/call_context.rs:59]: agreed.

File: src/call_context.rs
Comment (krl) [reply src/call_context.rs:121]: There's several places in the code that already panics where it shouldn't. I'll make a separate issue for this.

File: dusk-abi/src/lib.rs
Comment (krl) [reply dusk-abi/src/lib.rs:22]: Also related to visibility in the debug!() macro. wontfix

File: dusk-abi/src/lib.rs
Comment (krl) [reply dusk-abi/src/lib.rs:16]: Will also be deprecated by alloc.

File: dusk-abi/src/lib.rs
Comment (krl) [reply dusk-abi/src/lib.rs:5]: agreed.

**Review by ZER0 — APPROVED** [review]

**Pull request merged**
